### ISSUE TITLE
Add support for PostgreSQL database

### ISSUE THREAD

**Issue by php-coder** [php-coder/mystamps#1034]
- [x] adapt application to work with PostgreSQL
- [x] fix the migrations
- [x] review and fix the sequences
  - [x] users_id_seq
  - [x] series_id_seq
  - [x] collections_id_seq
  - [x] transaction_participants_id_seq
  - [x] series_sales_id_seq
  - [x] site_parsers_id_seq
  - [x] images
  - [x] *_catalog
- [x] review and ensure that we don't insert ids with prod profile
- [x] run integration tests on PostgreSQL as well (extracted into #1054)
- [x] add separate docker-compose file
- [x] update `README.md`/`NEWS.txt` files

**Comment by php-coder** [comment]
It doesn't start because one of migrations fails:

> 2019-06-06 01:32:03.476  WARN 1 --- [           main] ationConfigEmbeddedWebApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'liquibase' defined in class path resource [org/springframework/boot/autoconfigure/liquibase/LiquibaseAutoConfiguration$LiquibaseConfiguration.class]: Invocation of init method failed; nested exception is liquibase.exception.MigrationFailedException: Migration failed for change set classpath:/liquibase/version/0.3/2014-02-11--categories.xml::initiate-categories-table::php-coder:
     Reason: liquibase.exception.DatabaseException: ERROR: syntax error at or near "$"
  Position: 162 [Failed SQL: INSERT INTO public.categories (id, name, name_ru, created_at, created_by, updated_at, updated_by) VALUES ('1', 'Prehistoric animals', 'Доисторические животные', ${NOW}, '2', ${NOW}, '2')]

**Comment by php-coder** [comment]
UPDATE: this error happens only when "test-data" context is active in Liquibase (that means that the `test` and `travis` profiles should be affected by that).

**Comment by php-coder** [comment]
After fixing the previous issue (by declaring NOW property properly), the new issue appear:

> org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'liquibase' defined in class path resource [org/springframework/boot/autoconfigure/liquibase/LiquibaseAutoConfigura
tion$LiquibaseConfiguration.class]: Invocation of init method failed; nested exception is liquibase.exception.MigrationFailedException: Migration failed for change set classpath:/liquibase/version/0.3/2
014-05-28--release_month_and_day.xml::fill-release_year-column-based-on-values-from-released_at::php-coder:
     Reason: liquibase.exception.DatabaseException: ERROR: function year(date) does not exist
  Hint: No function matches the given name and argument types. You might need to add explicit type casts.
  Position: 35 [Failed SQL: UPDATE series
            SET release_year=YEAR(released_at)
            WHERE released_at IS NOT NULL]

**Comment by php-coder** [comment]
The next error is:

> Caused by: liquibase.exception.MigrationFailedException: Migration failed for change set classpath:/liquibase/version/0.3/2014-09-28--collection_slug.xml::update-collections-slug::php-coder:
     Reason: liquibase.exception.DatabaseException: ERROR: column "c" of relation "collections" does not exist
  Position: 29 [Failed SQL: UPDATE collections c
                       SET c.slug = (
                               SELECT LOWER(REPLACE(u.login, ' ', '-'))
                               FROM users u
                               WHERE u.id = c.user_id
                       )]

**Comment by php-coder** [comment]
The next error is:

> Caused by: liquibase.exception.MigrationFailedException: Migration failed for change set classpath:/liquibase/version/0.4/2015-06-22--image_url.xml::fill-series_images-table-based-on-values-from-image_url::php-coder:
      Reason: liquibase.exception.DatabaseException: ERROR: column "image_id" is of type integer but expression is of type text
Hint: You will need to rewrite or cast the expression.
Position: 75 [Failed SQL: INSERT INTO series_images(series_id, image_id)
                     SELECT id AS series_id, REPLACE(image_url, '/image/', '') AS image_id
                     FROM series]

**Comment by php-coder** [comment]
One more:

> Caused by: liquibase.exception.MigrationFailedException: Migration failed for change set classpath:/liquibase/version/0.4/2015-07-07--salt_and_hash.xml::increase-length-of-hash-column-in-users-table::php-coder:
      Reason: liquibase.exception.DatabaseException: ERROR: syntax error at or near "MODIFY"
   Position: 22 [Failed SQL: ALTER TABLE users
                        MODIFY COLUMN hash VARCHAR(60) NOT NULL]

**Comment by php-coder** [comment]
Yet another:

> Caused by: liquibase.exception.MigrationFailedException: Migration failed for change set classpath:/liquibase/version/0.4/2016-01-14--unique_slug_in_countries.xml::add-unique-constraint-to-slug-in-countries-table::AleksSPb:
     Reason:
          classpath:/liquibase/changelog.xml : liquibase.precondition.core.SqlPrecondition@74bdc168 : ERROR: column "num" does not exist
  Position: 72
...
Caused by: liquibase.exception.DatabaseException: Error executing SQL SELECT COUNT(*) as num
                               FROM countries
                               GROUP BY slug
                               HAVING num > 1
                               UNION SELECT 0: ERROR: column "num" does not exist

**Comment by php-coder** [comment]
The same but for categories:

> Caused by: liquibase.exception.MigrationFailedException: Migration failed for change set classpath:/liquibase/version/0.4/2016-07-22--unique_slug_in_categories.xml::add-unique-constraint-to-slug-in-categories-table::php-coder:
     Reason:
          classpath:/liquibase/changelog.xml : liquibase.precondition.core.SqlPrecondition@290b1b2e : ERROR: column "num" does not exist
  Position: 73
...
Caused by: liquibase.exception.PreconditionErrorException: Precondition Error
       at liquibase.precondition.core.SqlPrecondition.check(SqlPrecondition.java:64) ~[liquibase-core-3.5.5.jar!/:na]
       at liquibase.precondition.core.AndPrecondition.check(AndPrecondition.java:43) ~[liquibase-core-3.5.5.jar!/:na]
       at liquibase.precondition.core.PreconditionContainer.check(PreconditionContainer.java:202) ~[liquibase-core-3.5.5.jar!/:na]
       at liquibase.changelog.ChangeSet.execute(ChangeSet.java:498) ~[liquibase-core-3.5.5.jar!/:na]
       ... 34 common frames omitted
Caused by: liquibase.exception.DatabaseException: Error executing SQL SELECT COUNT(*) as num
                               FROM categories
                               GROUP BY slug
                               HAVING num > 1
                               UNION SELECT 0: ERROR: column "num" does not exist
  Position: 73

**Comment by php-coder** [comment]
Also there is another one in `0.4/2016-08-18--unique_slug_in_collections.xml::add-unique-constraint-to-slug-in-collections-table` but for collections.

**Comment by php-coder** [comment]
Now it fails with:

> Caused by: liquibase.exception.MigrationFailedException: Migration failed for change set classpath:/liquibase/version/0.4/2017-01-06--top_categories.xml::add-sport-category::php-coder:
     Reason: liquibase.exception.DatabaseException: ERROR: duplicate key value violates unique constraint "pk_categories"
  Detail: Key (id)=(1) already exists. [Failed SQL: INSERT INTO public.categories (name, name_ru, slug, created_at, created_by, updated_at, updated_by) VALUES ('Sport', '?????', 'sport', NOW(), (SELECT id FROM users WHERE role = 'ADMIN' ORDER by id LIMIT 1), NOW(), (SELECT id FROM users WHERE role = 'ADMIN' ORDER by id LIMIT 1))]

**Comment by php-coder** [comment]
One more:

> Caused by: liquibase.exception.MigrationFailedException: Migration failed for change set classpath:/liquibase/version/0.4/2017-05-29--test_image.xml::add-test-image::php-coder:
     Reason: liquibase.exception.DatabaseException: org.postgresql.util.PSQLException: ERROR: column "image_id" is of type integer but expression is of type character varying
  Hint: You will need to rewrite or cast the expression.
  Position: 58

**Comment by php-coder** [comment]
More errors, this time related to our usage of JdbcTemplate:

> org.springframework.dao.InvalidDataAccessApiUsageException: The getKey method should only be used when a single key is returned.  The current key entry contains multiple keys: [{id=4, name=test, created_at=2019-06-09 21:14:26.889, created_by=2, updated_at=2019-06-09 21:14:26.889, updated_by=2, name_ru=null, slug=test}]

**Comment by php-coder** [comment]
Migration is failing (from prod profile):

> Caused by: liquibase.exception.MigrationFailedException: Migration failed for change set classpath:/liquibase/version/0.4/2018-06-09--add_number_of_stamps_to_collections_series.xml::fill-collections_series-number_of_stamps-field::php-coder:
     Reason: liquibase.exception.DatabaseException: ERROR: column "cs" of relation "collections_series" does not exist
  Position: 37 [Failed SQL: UPDATE collections_series cs
                       SET cs.number_of_stamps = (
                               SELECT s.quantity
                               FROM series s
                               WHERE s.id = cs.series_id
                       )]

**Comment by php-coder** [comment]
We need to check the following sequences to ensure that their next values haven't been already used by migrations with hardcoded IDs:

```
mystamps=# select sequence_name from information_schema.sequences;
             sequence_name
---------------------------------------
 scott_catalog_id_seq
 users_id_seq
 countries_id_seq
 series_id_seq
 michel_catalog_id_seq
 yvert_catalog_id_seq
 gibbons_catalog_id_seq
 images_id_seq
 images_data_id_seq
 suspicious_activities_types_id_seq
 suspicious_activities_id_seq
 categories_id_seq
 collections_id_seq
 series_import_requests_id_seq
 transaction_participants_id_seq
 series_sales_id_seq
 top_categories_id_seq
 series_import_request_statuses_id_seq
 transaction_participant_groups_id_seq
 solovyov_catalog_id_seq
 zagorski_catalog_id_seq
 site_parsers_id_seq
```

**Comment by 0pdd** [comment]
@php-coder the puzzle [#1039](https://github.com/php-coder/mystamps/issues/1039) is still not solved.

**Comment by php-coder** [comment]
The following sequences have to be fixed:
- users_id_seq
- series_id_seq
- collections_id_seq
- transaction_participants_id_seq
- series_sales_id_seq
- site_parsers_id_seq

Here is how to test that it's OK:
```
mystamps=# SELECT 'zagorski_catalog_id_seq', (SELECT MAX(id) FROM zagorski_catalog), last_value FROM zagorski_catalog_id_seq;
        ?column?         | max | last_value
-------------------------+-----+------------
 zagorski_catalog_id_seq |   1 |          1
(1 row)
```
The `max` and `last_value` must have the same value.

Here how it looks when it doesn't work:
```
mystamps=# SELECT 'users_id_seq', (SELECT MAX(id) FROM users), last_value FROM users_id_seq;
   ?column?   | max | last_value
--------------+-----+------------
 users_id_seq |   5 |          1
(1 row)
mystamps=# insert into users(login,role,name,email,hash,registered_at,activated_at) values('test','USER','TEST','test@mail.ru','XXX',NOW(),NOW());
ERROR:  duplicate key value violates unique constraint "pk_users"
DETAIL:  Key (id)=(1) already exists.
```

**Comment by php-coder** [comment]
Some of the fixes broke change checksums check:

```
     5 change sets check sum
          classpath:/liquibase/version/0.3/2014-05-28--release_month_and_day.xml::fill-release_year-column-based-on-values-from-released_at::php-coder was: 7:9d233932224649aa730fd82df77f22a4 but is now: 7:809596648b6e7ef5de74d5c690611f53
          classpath:/liquibase/version/0.3/2014-09-28--collection_slug.xml::update-collections-slug::php-coder was: 7:e097ed4a1bcc0a584aa5b385cff2d03a but is now: 7:8af814be468761d310c0a46c5f5cfbbf
          classpath:/liquibase/version/0.4/2015-06-22--image_url.xml::fill-series_images-table-based-on-values-from-image_url::php-coder was: 7:b0d069aacd68b75cacdc22a6c0fb9503 but is now: 7:abff5f789b843a3131a331f9e99ce92c
          classpath:/liquibase/version/0.4/2015-07-07--salt_and_hash.xml::increase-length-of-hash-column-in-users-table::php-coder was: 7:4160c036a7d7c4cae76ce701d8c4e52a but is now: 7:85d6c91356b5832a281669e7ec0e217c
          classpath:/liquibase/version/0.4/2018-06-09--add_number_of_stamps_to_collections_series.xml::fill-collections_series-number_of_stamps-field::php-coder was: 7:acf5541cb113870c46be460417935cf9 but is now: 7:1333048e062daba328f7af56a3ec9c61
```

**Comment by 0pdd** [comment]
@php-coder 2 puzzles [#1039](https://github.com/php-coder/mystamps/issues/1039), [#1054](https://github.com/php-coder/mystamps/issues/1054) are still not solved.

**Comment by php-coder** [comment]
All possible duplicate key violations should be fixed now. In order to detect them, I used the following script:

```console
$ cd docker
$ docker-compose -f docker-compose.yml -f postgres.yml down -v
$ docker-compose -f docker-compose.yml -f postgres.yml up -d
$ docker-compose -f docker-compose.yml -f postgres.yml logs -f
$ docker cp test-insert.sql mystamps_db_1:/
$ docker exec -it mystamps_db_1 psql -U mystamps -d mystamps -W -f test-insert.sql
```

and here is the content of `test-insert.sql` file:

```sql
INSERT INTO categories(name,created_at,created_by,updated_at,updated_by,slug)
VALUES('Test',now(),1,now(),1,'test');
INSERT INTO categories_aliases(category_id,name)
VALUES(1,'Foo');
INSERT INTO collections(user_id,slug,updated_at,updated_by)
VALUES(1,'test',now(),1);
INSERT INTO countries(name,created_at,created_by,updated_at,updated_by,slug)
VALUES('Test',now(),1,now(),1,'test');
INSERT INTO countries_aliases(country_id,name)
VALUES(1,'Bar');
INSERT INTO gibbons_catalog(code)
VALUES('GIB');
INSERT INTO images(type)
VALUES('PNG');
INSERT INTO images_data(content,image_id,preview)
SELECT decode('DEADBEEF', 'hex'), 1, TRUE;
INSERT INTO michel_catalog(code)
VALUES('MIC');
INSERT INTO scott_catalog(code)
VALUES('SCO');
INSERT INTO series(quantity,perforated,created_at,created_by,updated_at,updated_by,category_id)
VALUES(10,TRUE,now(),1,now(),1,1);
INSERT INTO collections_series(collection_id,series_id,number_of_stamps)
VALUES(1,1,10);
INSERT INTO series_gibbons_catalog(series_id,gibbons_id)
VALUES(1,1);
INSERT INTO series_images(series_id,image_id)
VALUES(1,1);
INSERT INTO series_import_requests(url,status_id,requested_at,requested_by,updated_at)
VALUES('http://example.com',1,now(),1,now());
INSERT INTO series_import_parsed_data(request_id,created_at,updated_at)
VALUES(1,now(),now());
INSERT INTO series_import_raw_data(request_id,page_content,created_at,updated_at)
VALUES(1,'<html>TEST</html>',now(),now());
INSERT INTO series_import_request_statuses(name)
VALUES('New status');
INSERT INTO series_michel_catalog(series_id,michel_id)
VALUES(1,1);
INSERT INTO series_sales_import_parsed_data(request_id,created_at,updated_at)
VALUES(1,now(),now());
INSERT INTO series_scott_catalog(series_id,scott_id)
VALUES(1,1);
INSERT INTO solovyov_catalog(code)
VALUES('SOL');
INSERT INTO series_solovyov_catalog(series_id,solovyov_id)
SELECT 1, id FROM solovyov_catalog WHERE code = 'SOL';
INSERT INTO yvert_catalog(code)
VALUES('YVT');
INSERT INTO series_yvert_catalog(series_id,yvert_id)
VALUES(1,1);
INSERT INTO zagorski_catalog(code)
VALUES('ZAG');
INSERT INTO series_zagorski_catalog(series_id,zagorski_id)
SELECT 1, id FROM zagorski_catalog WHERE code = 'ZAG';
INSERT INTO similar_series(series_id,similar_series_id)
VALUES(1,1);
INSERT INTO site_parsers(name)
VALUES('test parser');
INSERT INTO site_parser_params(parser_id,name,value)
VALUES(1,'foo','bar');
INSERT INTO suspicious_activities(page,type_id,ip,occurred_at)
VALUES('/foo',1,'127.0.0.1',now());
INSERT INTO suspicious_activities_types(name)
VALUES('weird');
INSERT INTO top_categories(name,created_at,created_by,updated_at,updated_by)
VALUES('parent',now(),1,now(),1);
INSERT INTO transaction_participant_groups(name,created_at,created_by,updated_at,updated_by)
VALUES('group',now(),1,now(),1);
INSERT INTO transaction_participants(name,is_buyer,is_seller)
VALUES('seller',FALSE,TRUE);
INSERT INTO series_sales(series_id,seller_id,first_price,first_currency,created_at,created_by)
VALUES(1,1,10,'RUB',now(),1);
INSERT INTO users(login,role,name,email,hash,registered_at,activated_at)
VALUES('login','USER','Name','login@example.com','XXX',now(),now());
INSERT INTO users_activation(act_key,email,created_at,lang)
VALUES('1234','login@example.com',now(),'ru');
```

**Comment by php-coder** [comment]
The application starts with PostgreSQL on test/prod profiles. The issue can be closed now.

**Comment by 0pdd** [comment]
@php-coder 9 puzzles [#1039](https://github.com/php-coder/mystamps/issues/1039), [#1134](https://github.com/php-coder/mystamps/issues/1134), [#1135](https://github.com/php-coder/mystamps/issues/1135), [#1136](https://github.com/php-coder/mystamps/issues/1136), [#1137](https://github.com/php-coder/mystamps/issues/1137), [#1138](https://github.com/php-coder/mystamps/issues/1138), [#1139](https://github.com/php-coder/mystamps/issues/1139), [#1140](https://github.com/php-coder/mystamps/issues/1140), [#1141](https://github.com/php-coder/mystamps/issues/1141) are still not solved; solved: [#1054](https://github.com/php-coder/mystamps/issues/1054).

**Comment by 0pdd** [comment]
@php-coder 8 puzzles [#1039](https://github.com/php-coder/mystamps/issues/1039), [#1134](https://github.com/php-coder/mystamps/issues/1134), [#1135](https://github.com/php-coder/mystamps/issues/1135), [#1136](https://github.com/php-coder/mystamps/issues/1136), [#1137](https://github.com/php-coder/mystamps/issues/1137), [#1138](https://github.com/php-coder/mystamps/issues/1138), [#1139](https://github.com/php-coder/mystamps/issues/1139), [#1141](https://github.com/php-coder/mystamps/issues/1141) are still not solved; solved: [#1054](https://github.com/php-coder/mystamps/issues/1054), [#1140](https://github.com/php-coder/mystamps/issues/1140).
